### PR TITLE
core: namespace module objects and function extensions

### PR DESCRIPTION
**Pull request by sipsma**
This change results in names being prefixed by module name for:
1. Any objects defined by the module
2. Any functions defined by the module that are extending objects from other pre-existing modules (or core)

This reduces the chance of conflicts when depending on 2+ modules that happen to use overlapping names.

---

Bringing out of draft:
* cleaned up casing+namespacing consistency (see EDIT in [this comment](https://github.com/dagger/dagger/pull/5932#issuecomment-1771787451))
* made minimal changes to docs so they still work. Further docs on namespacing will be part of our upcoming overall effort to improve zenith documentation.

### CODE REVIEW

**Comment by sipsma** [comment]
In noticed some discrepancies in how we handle casing in various places while doing this. The current code in this PR handles it but it's not very solid ground, especially as more SDKs get added and/or modified.

For object names:
1. In the schema, it's CamelCase
2. In Go SDK's codegen'd `invoke` func, it's CamelCase
3. In Python SDK `_call`, I *think* it might be the literal module name unchanged (is that correct @helderco? didn't find the exact line in the python sdk to confirm)

For field names:
1. In the schema, it's lowerCamelCase
2. In `Module.currentFunctionCall.Name`, it's CamelCase
3. In Go SDK's codegen'd `invoke` func, it's CamelCase
4. In Python SDK, I *think* it's CamelCase too? Again not very sure

---

It's probably better to just squash this as part of this PR.

I'm gonna give a shot at seeing if it's possible to always provide casing that matches whatever the SDK makes its calls with (e.g. if Python sets a function typedef's name to be `snake_case`, we give it it back in that casing).

That would be ideal because it would result in the least work for SDK devs. If it's not possible for some reason, I'll just make sure the casing expected is well defined and aligned everywhere.

**Comment by sipsma** [comment]
Also appended a technically unrelated commit that fixes some issues w/ cache mounts in rust tests: https://github.com/dagger/dagger/pull/5932/commits/510e4edff51092f84efacdf04f8290e1161536c7

It is most likely caused by the fact that we just upgraded to exclusively using dagger runners for CI. Easier to just include here for now, but will split out to separate PR if needed.

**Comment by helderco** [comment]
> For object names:
> [...]
> 3. In Python SDK `_call`, I _think_ it might be the literal module name unchanged (is that correct @helderco? didn't find the exact line in the python sdk to confirm)

There's no support for classes yet, so top level functions are registered as part of an API object with the name of the module, casing unchanged:

https://github.com/dagger/dagger/blob/c213feb051f5ba8c270fd21814015bfc6363c0d5/sdk/python/src/dagger/ext/_module.py#L78-L80

I didn't change the casing because it seemed the API was doing that already.

File: internal/mage/sdk/rust.go [hunk 1]
```diff
@@ -227,7 +227,7 @@ func (Rust) rustBase(ctx context.Context, c *dagger.Client, image string) *dagge
 	base := c.
 		Container().
 		From(image).
-		WithMountedCache("~/.cargo", c.CacheVolume("rust-cargo")).
+		WithMountedCache("~/.cargo", c.CacheVolume("rust-cargo-"+image)).
```
Comment (vito) [inline internal/mage/sdk/rust.go:230]: Huh, mildly surprised that using `~` works here. Wouldn't we need to explicitly expand that somewhere? From the code I don't see that happening, so I wonder if this is mounting to literally `./~/.cargo` and it's working without the cache atm.

Comment (sipsma) [reply internal/mage/sdk/rust.go:230]: Oh it probably doesn't work, good catch, I'll just fix that too while I'm modifying this

**Comment by sipsma** [comment]
> There's no support for classes yet, so top level functions are registered as part of an API object with the name of the module, casing unchanged:
> 
> https://github.com/dagger/dagger/blob/c213feb051f5ba8c270fd21814015bfc6363c0d5/sdk/python/src/dagger/ext/_module.py#L78-L80
> 
> I didn't change the casing because it seemed the API was doing that already.

Okay cool. The code in this PR now just keeps track of both the "normalized" name of each object/function (i.e. the name + casing used in the graphql API) and the "original" name they were registered with by the SDK. When invoking the SDK runtime, we use the original name; everywhere else uses the normalized name. So SDK Modules can feel free to use whatever casing they want and shouldn't need to try to handle casing/namespacing changes.

**Comment by sipsma** [comment]
> My understanding is some chunk of this is immediately undone by https://github.com/dagger/dagger/pull/5954 - but it makes sense to do this in two steps anyhow.

Yep exactly, it only undoes the namespacing of fields, which no longer applies, so just a matter of a few lines of code

**Pull request merged**